tedana/metrics/collect.py

```diff
@@ -0,0 +1,225 @@
+        Mixing matrix after sign flipping and sorting.
+    """
+    if metrics is None:
+        metrics = ['map weight']
+    RepLGR.info('The following metrics were calculated: {}.'.format(', '.join(metrics)))
+
+    if not (data_cat.shape[0] == data_optcom.shape[0] == mask.shape[0]):
+        raise ValueError('First dimensions (number of samples) of data_cat ({0}), '
+                         'data_optcom ({1}), and mask ({2}) do not '
+                         'match'.format(data_cat.shape[0], data_optcom.shape[0],
+                                        mask.shape[0]))
+    elif data_cat.shape[1] != len(tes):
+        raise ValueError('Second dimension of data_cat ({0}) does not match '
+                         'number of echoes provided (tes; '
+                         '{1})'.format(data_cat.shape[1], len(tes)))
+    elif not (data_cat.shape[2] == data_optcom.shape[1] == mixing.shape[0]):
+        raise ValueError('Number of volumes in data_cat ({0}), '
+                         'data_optcom ({1}), and mixing ({2}) do not '
+                         'match.'.format(data_cat.shape[2], data_optcom.shape[1], mixing.shape[0]))
+
+    INPUTS = ['data_cat', 'data_optcom', 'mixing', 'mask', 'tes', 'ref_img']
+    METRIC_DEPENDENCIES = {
+        'kappa': ['map FT2', 'map Z'],
+        'rho': ['map FS0', 'map Z'],
+        'countnoise': ['map Z', 'map Z clusterized'],
+        'countsigFT2': ['map FT2 clusterized'],
+        'countsigFS0': ['map FS0 clusterized'],
+        'dice_FT2': ['map beta T2 clusterized', 'map FT2 clusterized'],
+        'dice_FS0': ['map beta S0 clusterized', 'map FS0 clusterized'],
+        'signal-noise_t': ['map Z', 'map Z clusterized', 'map FT2'],
+        'variance explained': ['map optcom betas'],
+        'normalized variance explained': ['map weight'],
+        'd_table_score': ['kappa', 'dice_FT2', 'signal-noise_t',
+                          'countnoise', 'countsigFT2'],
+        'map FT2': ['map Z', 'data_cat', 'mask'],
+        'map FS0': ['map Z'],
+        'map Z': ['map weight'],
+        'map weight': ['data_optcom', 'mixing'],
+        'map optcom betas': ['data_optcom', 'mixing'],
+        'map percent signal change': ['data_optcom', 'map optcom betas'],
+        'map Z clusterized': ['map Z', 'mask', 'ref_img', 'tes'],
+        'map FT2 clusterized': ['map FT2', 'mask', 'ref_img', 'tes'],
+        'map FS0 clusterized': ['map FS0', 'mask', 'ref_img', 'tes'],
+        'map beta T2 clusterized': ['map FT2 clusterized', 'map optcom betas',
+                                    'countsigFT2', 'mask', 'ref_img', 'tes'],
+        'map beta S0 clusterized': ['map FS0 clusterized', 'map optcom betas',
+                                    'countsigFS0', 'mask', 'ref_img', 'tes'],
+    }
+    # Apply masks before anything else
+    data_cat = data_cat[mask, ...]
+    data_optcom = data_optcom[mask, :]
+
+    required_metrics = dependency_resolver(METRIC_DEPENDENCIES, metrics, INPUTS)
```

Comment:
Just noting that required_metrics is unordered, but by having an ordered list of if statements for each metric calculation, they'll always be calculated after dependencies. This approach might get bloated if the metric list keeps growing, but this should be fine for now.